log4j-core/src/main/java/org/apache/logging/log4j/core/impl/Log4jLogEvent.java

```diff
@@ -43,94 +43,117 @@
-    // Location data
-    private final String loggerFqcn;
-    private @Nullable StackTraceElement source;
+    // This field is mutable, but its state is not shared with other objects.
+    private final MutableInstant instant = new MutableInstant();
+
+    // 2. Fields with setters, initialized in the constructor.
+    private boolean endOfBatch;
     private boolean includeLocation;
 
+    // 3. Fields with an immutable type, initialized lazily.
+    //    These fields self-initialize if not provided.
+    private @Nullable StackTraceElement source;
+    private String threadName;
+    private long threadId;
+    private int threadPriority;
+
+    // 4. Fields with a potentially mutable type.
+    //    These fields can cause mutability problems for Log4jLogEvent.
+    private Message message;
+    private final StringMap contextData;
+    private final ThreadContext.ContextStack contextStack;
+
     /** LogEvent Builder helper class. */
     public static class Builder implements org.apache.logging.log4j.plugins.util.Builder<LogEvent> {
 
+        // 1. Fields with an immutable type, initialized eagerly.
+        //    These fields always keep the value assigned.
         private String loggerFqcn;
-        private Marker marker;
         private Level level;
         private String loggerName;
-        private Message message;
+        private Marker marker;
         private Throwable thrown;
+        private boolean endOfBatch;
+        private boolean includeLocation;
+        private long nanoTime;
+        // This field is mutable, but it is always copied.
         private final MutableInstant instant = new MutableInstant();
-        private StringMap contextData;
-        private ThreadContext.ContextStack contextStack = ThreadContext.getImmutableStack();
-        private long threadId;
+
+        // 2. Fields with an immutable type, initialized lazily.
+        //    These fields self-initialize if not provided.
+        private StackTraceElement source;
         private String threadName;
+        private long threadId;
         private int threadPriority;
-        private StackTraceElement source;
-        private boolean includeLocation;
-        private boolean endOfBatch = false;
-        private long nanoTime;
+
+        // 3. Fields with a mutable type.
+        //    These fields require special handling.
+        private Message message;
+        private StringMap contextData;
+        private ThreadContext.ContextStack contextStack;
+
+        // 4. Fields with dependency-injected values.
         private Clock clock;
         private ContextDataInjector contextDataInjector;
 
         public Builder() {
             initDefaultContextData();
         }
 
+        /**
+         * Initializes the builder with an <strong>immutable</strong> instance or a copy of the log event fields.
+         *
+         * @param other The log event to copy.
+         */
         public Builder(final LogEvent other) {
```

Comment:
_Nitpick_: instead of a public constructor, we could have a `newBuilder(LogEvent)` factory method.